agent/grpc-external/services/resource/write.go

```diff
@@ -0,0 +1,144 @@
+package resource
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid/v2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/lib/retry"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (s *Server) Write(ctx context.Context, req *pbresource.WriteRequest) (*pbresource.WriteResponse, error) {
+	if err := validateWriteRequiredFields(req); err != nil {
+		return nil, err
+	}
+
+	reg, err := s.resolveType(req.Resource.Id.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the user sent the correct type of data.
+	if !req.Resource.Data.MessageIs(reg.Proto) {
+		got := strings.TrimPrefix(req.Resource.Data.TypeUrl, "type.googleapis.com/")
+
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"resource.data is of wrong type (expected=%q, got=%q)",
+			reg.Proto.ProtoReflect().Descriptor().FullName(),
+			got,
+		)
+	}
+
+	var result *pbresource.Resource
+	err = s.retryCAS(ctx, req.Resource.Version, func() error {
+		input := clone(req.Resource)
+
+		// Read with EventualConsistency because we'll already automatically retry on
+		// CAS failure, so it's not worth the latency penalty of StrongConsistency.
+		existing, err := s.Backend.Read(ctx, storage.EventualConsistency, input.Id)
```

Comment:
Thanks for the call out! To clarify, the whole retry-loop thing is less of an optimization, and more to hide the fact we're doing a CAS operation from the user.

For context, all writes have to be CAS operations if we want to (safely) do things like:
- Ensure the `Uid` doesn't change between writes.
- Copy the resource `Status` between writes.
- Pass the current version of a resource to pre-write validation/mutation hooks.

That is, unless we only support more feature-rich storage backends that allow you to do "patch"-style updates.

Generally, people don't need CAS semantics in the same way that scripts and controllers do, they just want to overwrite whatever's currently there. So it'd be confusing for them to see a message about a CAS operation they didn't know was happening having failed because of a concurrent write 😄 hence the retry-loop.

My thinking regarding reading with `EventualConsistency` was that because of our fairly conservative network latency recommendations, Raft followers are usually less than a few hundred milliseconds behind the leader in a healthy cluster. In the majority of cases, the first read will return the latest version, so it seems like a reasonable trade-off to avoid having to do 2 round trips to the followers (one to verify leadership for the consistent read, and one to append the write log).

Does that sound ok? If so I'll update the comments to be clearer.